src/librustc_mir/transform/simplify.rs

```diff
@@ -382,6 +382,11 @@ impl<'a, 'tcx> Visitor<'tcx> for DeclMarker<'a, 'tcx> {
                             }
                         }
                     }
+                } else if let StatementKind::Assign(box (p, Rvalue::Discriminant(d))) = &stmt.kind {
+                    if !p.is_indirect() {
+                        trace!("skipping store of discriminant value {:?} to {:?}", d, p);
```

Comment:
This confused me at first, but basically this is treating certain initializers as not keeping the destination alive, right?
Something very important in these checks though: you need to make sure `p.local` is `local` otherwise it may be somewhere else in the statement (you shouldn't trust `ctx` to let you know where it is, IMO).

Also, it would be nice if `if let StatementKind::Assign(box (p, ` was only once, and then the only check for the `Rvalue` was whether we consider it to have side-effects.

Also, does this pass re-run itself to fixpoint? Because if you remove an assignment to a dead local, you remove uses of other locals also mentioned in that assignment, which may be their last.

You can't do this in one pass because e.g. in `x = y`, `y` might be unused *only* if `x` is unused, and you can't know  that ahead of time.

Reply:
> basically this is treating certain initializers as not keeping the destination alive, right?

Yup, that's correct.

> Something very important in these checks though: you need to make sure p.local is local otherwise it may be somewhere else in the statement (you shouldn't trust ctx to let you know where it is, IMO).

I'm struggling to think of a case where that would happen. Do you have an example in mind? 

> Also, does this pass re-run itself to fixpoint? Because if you remove an assignment to a dead local, you remove uses of other locals also mentioned in that assignment, which may be their last.

It doesn't currently. That would be a good improvement though.

Reply:
> I'm struggling to think of a case where that would happen. Do you have an example in mind?

The point is to check instead of assuming, especially in this case where it's basically free.
If nothing else it serves as self-documenting code.